solver/src/main/java/org/chocosolver/solver/constraints/binary/PropSquare.java

```diff
@@ -117,57 +108,95 @@ protected void updateLowerBoundofX() throws ContradictionException {
 
     protected void updateUpperBoundofX() throws ContradictionException {
         vars[0].updateUpperBound(Math.max(sqr(vars[1].getLB()), sqr(vars[1].getUB())), this);
-
     }
 
     protected boolean updateHolesinX() throws ContradictionException {
+        if (!vars[0].hasEnumeratedDomain()) {
+            return false;
+        }
+        boolean impact = false;
+        boolean isFirstPropagation = model.getSolver().getNodeCount() == 0;
+        if (isFirstPropagation) {
+            // check perfect squares once and for all
+            int ub = vars[0].getUB();
+            vrms.clear();
+            vrms.setOffset(vars[0].getLB());
+            for (int value = vars[0].getLB(); value <= ub; value = vars[0].nextValue(value)) {
+                if (!MathUtils.isPerfectSquare(value)) {
+                    vrms.add(value);
+                }
+            }
+            impact = vars[0].removeValues(vrms, this);
+        }
+
         // remove intervals to deal with consecutive value removal and upper bound modification
         if (bothEnum) {
             int ub = vars[0].getUB();
             vrms.clear();
             vrms.setOffset(vars[0].getLB());
             for (int value = vars[0].getLB(); value <= ub; value = vars[0].nextValue(value)) {
-                if (!(MathUtils.isPerfectSquare(value) &&
-                        (vars[1].contains(floor_sqrt(value)) || vars[1].contains(-floor_sqrt(value))))) {
+                int sqrt = floor_sqrt(value);
+                if (!vars[1].contains(sqrt) && !vars[1].contains(-sqrt)) {
                     vrms.add(value);
                 }
             }
-            return vars[0].removeValues(vrms, this);
-        } else if (vars[0].hasEnumeratedDomain()) {
+            impact |= vars[0].removeValues(vrms, this);
+        } else {
             int value = vars[0].getLB();
             int nlb = value - 1;
             while (nlb == value - 1) {
-                if (!vars[1].contains(floor_sqrt(value)) && !vars[1].contains(-floor_sqrt(value))) {
+                int sqrt = floor_sqrt(value);
```

Comment:
This statement is valid if and only if value is a perfect square value (which is valid if and only if vars[0] has an enumerated domain). Otherwise, it should be `int sqrt = ceil_sqrt(value);`.